crates/oxc_linter/src/rules/jest/no_done_callback.rs

```diff
@@ -0,0 +1,308 @@
+use oxc_ast::{
+    ast::{Argument, CallExpression, Expression, FormalParameters},
+    AstKind,
+};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+use crate::{
+    context::LintContext,
+    jest_ast_util::{
+        get_node_name, parse_general_jest_fn_call, JestFnKind, JestGeneralFnKind,
+        ParsedGeneralJestFnCall,
+    },
+    rule::Rule,
+    AstNode,
+};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint(jest/no-done-callback): {0:?}")]
+#[diagnostic(severity(warning), help("{1:?}"))]
+struct NoDoneCallbackDiagnostic(&'static str, &'static str, #[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoDoneCallback;
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// This rule checks the function parameter of hooks & tests for use of the done argument, suggesting you return a promise instead.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// When calling asynchronous code in hooks and tests, jest needs to know when the asynchronous work is complete to progress the current run.
+    /// Originally the most common pattern to achieve this was to use callbacks:
+    ///
+    /// ```javascript
+    /// test('the data is peanut butter', done => {
+    ///   function callback(data) {
+    ///     try {
+    ///       expect(data).toBe('peanut butter');
+    ///       done();
+    ///     } catch (error) {
+    ///       done(error);
+    ///     }
+    ///   }
+    ///
+    ///   fetchData(callback);
+    /// });
+    /// ```
+    ///
+    /// This can be very error-prone however, as it requires careful understanding of how assertions work in tests or otherwise tests won't behave as expected.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// beforeEach(done => {
+    ///   // ...
+    /// });
+    ///
+    /// test('myFunction()', done => {
+    ///   // ...
+    /// });
+    ///
+    /// test('myFunction()', function (done) {
+    ///   // ...
+    /// });
+    /// ```
+    NoDoneCallback,
+    // TODO: add suggestion (see jest-community/eslint-plugin-jest#586)
+    restriction
```

Comment:
Not implement its auto-suggestion, is this suitable?

Reply:
I thought about having a `--jest-plugin` flag for turning on jest rules, so we can put these rules in the right place. But let's figure this out later.